content/issue-2/how-i-build-backon.md

```diff
@@ -0,0 +1,191 @@
+
+## Design
+
+My first idea was to use an `Iterator<Item = Duration>` to represent the backoff. Any iterator that can return a `Duration` type can be used as backoff. Using an iterator to represent backoff has a very direct and clear meaning, and users can easily understand and implement it without reading every function's comments. Secondly, I wanted to provide a usage experience for backoff similar to Rust's native functions:
+
+```rust
+async fn fetch_url(url: &str) -> Result<String, reqwest::Error> {
+  	fetch.retry(ExponentialBackoff::default()).await
+}
+```
+
+It looks great: simple and direct, doesn't disrupt the user's reading order, and allows the user to locate the business logic position at a glance. Let's get started implementing it!
+
+## Implementation
+
+First of all, what we need to understand is that async functions in Rust are essentially generators. These generators capture variables from the current environment and generate an anonymous Future. To retry an async function, we need to call the generator again to generate a brand new Future to execute.
+
+I once went down the wrong path with a failed demo for retry: [we can't retry a future directly](https://github.com/Xuanwo/backon/pull/1). At that time, I naively tried to retry a TryFuture directly:
+
+```rust
+pub trait Retryable<B: Policy, F: Fn(&Self::Error) -> bool>: TryFuture + Sized {
+    fn retry(self, backoff: B, handle: F) -> Retry<Self, B, F>;
+}
+```
+
+Now I understand that this approach is incorrect. Once a `Future` enters the `Poll::Ready` state, we should not poll it again, as documented:
+
+> Once a future has completed (returned `Ready` from `poll`), calling its `poll` method again may panic, block forever, or cause other kinds of problems
+
+Next, I needed to adjust my thinking and focus on implementing for `|| -> impl Future<Result<T>>`. First, I defined a `Retryable` trait and implemented it for all `FnMut() -> Fut` types:
+
+```rust
+pub trait Retryable<
+    B: BackoffBuilder,
+    T,
+    E,
+    Fut: Future<Output = Result<T, E>>,
+    FutureFn: FnMut() -> Fut,
+>
+{
+    /// Generate a new retry
+    fn retry(self, builder: &B) -> Retry<B::Backoff, T, E, Fut, FutureFn>;
+}
+
+impl<B, T, E, Fut, FutureFn> Retryable<B, T, E, Fut, FutureFn> for FutureFn
+where
+    B: BackoffBuilder,
+    Fut: Future<Output = Result<T, E>>,
+    FutureFn: FnMut() -> Fut,
+{
+    fn retry(self, builder: &B) -> Retry<B::Backoff, T, E, Fut, FutureFn> {
+        Retry::new(self, builder.build())
+    }
+}
+```
+
+This trait involves the following type parameters:
+
+- `B: BackoffBuilder`: the backoff builder passed in by the user, which specifies different backoff parameters
+- `FutureFn: FnMut() -> Fut`: indicates that its type is a function that returns a `Fut`
+  - `FnOnce` requires taking ownership and cannot be called multiple times
+  - `Fn` can only get a `&self` reference, which can be limited in many scenarios
```

Comment:
For example, a sentence similar to "why the trait bound for FutureFn is FunMut() -> Fut" is necessary.

Reply:
I prefer to remove them since they are implementation details of backon that are not related to the topic of this article. Perhaps we could create a separate article to discuss the differences between `Fn`, `FnMut`, and `FnOnce`.